Fix issue #21301: add subject interest when focus is moved from subjectInterest field.

JayanthK007 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #21301.
2. This PR does the following: [When user enters the subject Interest and move the focus to other field it saves, before I was not saving until we press enter key.]
3. (For bug-fixing PRs only) The original bug occurred because: [The bug was caused because before subject field was saving only when enter key is pressed and even E2E testing was added such that enter key is pressed. This PR adds subject interest even if the focus is changed to any other field. The PR which caused this issue is #13219]

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [ x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x ] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [ x] I have written tests for my code.
- [ x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [ x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).


<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

## Proof that changes are correct

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

before changes

https://github.com/user-attachments/assets/e0705552-21b2-4e28-948b-095c2c1651f0

after changes


https://github.com/user-attachments/assets/3758c261-593f-4d77-9721-9206c81a1641

PTAL @seanlip 

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

JayanthK007 committed:
subject interest input takes values even if enter key is not pressed and even test cases for E2E testing are added

JayanthK007 commented:
I have added in testing when we press save button to check for changing focus. But we can add `.oppia-page-card` selector as it imitates user action after entering subject interest field.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/logged-in-user.ts:
```diff
@@ -789,18 +789,33 @@ export class LoggedInUser extends BaseUser {
 
+  /**
+   * Updates the user's subject interests in preference page.
+   * @param {string[]} interests - The new interests to set for the user when focus is moved to the next input or anywhere in the page.
```
"... on the page"

seanlip replied:
@JayanthK007 Please address this comment, thanks.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/logged-in-user.ts:
```diff
@@ -789,18 +789,33 @@ export class LoggedInUser extends BaseUser {
+      await this.page.click(subjectInterestsInputSelector); // Refocus the input field
+      await this.type(subjectInterestsInputSelector, interest);
+      // Click on the save button to trigger onBlur
+      await this.page.click('.e2e-test-save-changes-button');
```
Save '.e2e-test-save-changes-button' as a constant in the beginning of this file.

imchristie commented on core/tests/puppeteer-acceptance-tests/utilities/user/logged-in-user.ts:
```diff
@@ -789,18 +789,33 @@ export class LoggedInUser extends BaseUser {
   /**
    * Updates the user's subject interests in preference page.
-   * @param {string[]} interests - The new interests to set for the user.
+   * @param {string[]} interests - The new interests to set for the user when enter key is pressed.
```
"when"? Do you mean "and press the enter key"? Ditto with the function below.

imchristie reviewed (commented):
Reviewed. Please also fix the lint checks and other CI checks.

JayanthK007 committed:
added save button as constant and also lint checked

JayanthK007 committed:
all changes made

JayanthK007 committed:
removed lint error file and corrected it

JayanthK007 commented:
Hi @imchristie. 
I have made code changes and removed unwanted comments. Can you please check again?

JayanthK007 committed:
added focus to update where ever we touch in body tag

JayanthK007 committed:
Update logged-in-user.ts

removed click behaviour

JayanthK007 committed:
added focus for updateSubjectInterestWithBlurEvent

JayanthK007 commented:
hi @Nik-09 @imchristie . Test cases are passing for web and it is only failing for mobile testing. I don't know why. Even frontend Karma tests are failing for taking long time. Only these are failing. Can you guys please help me how to resolve this.

seanlip commented on oppia:
Please get rid of this change, thanks. (You can look at the "Files Changed" tab to verify whether it has gone or not.)

JayanthK007 replied:
Hi @seanlip i did'nt get you. Can you please tell me what changes i need to get rid of?

seanlip replied:
This change at the bottom of your "Files Changed" tab should not be there. (You'll see that it's absent in other PRs, too.)

![Screenshot from 2024-12-13 22-23-47](https://github.com/user-attachments/assets/04c4e260-89b8-4db6-95b1-b71804dddb6c)

JayanthK007 replied:
sure @seanlip i'll remove it.

JayanthK007 replied:
Hi @seanlip I have removed the oppia folder.

JayanthK007 reviewed (commented):
reviewed all changes

seanlip committed:
Merge branch 'develop' into subject-interest

Nik-09 commented:
Hi @JayanthK007, I will take a look into this PR by tomorrow.
Can you please respond to [this comment](https://github.com/oppia/oppia/pull/21389#issuecomment-2533571197) by @imchristie.

Thanks!

Nik-09 reviewed (approved):
LGTM

JayanthK007 commented:
> @JayanthK007 Please address every comment by commenting "Done" if it's fixed/addressed. After commenting, finish your review by clicking the "Submit Review" in the file changed section. For the CI checks, did you run the tests locally and did they pass locally?

I'm running CI checks but some are failing not getting why it is failing.

JayanthK007 committed:
Merge branch 'subject-interest' of https://github.com/JayanthK007/oppia into develop

seanlip commented:
@JayanthK007 Please follow these instructions if your CI checks fail: https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR

JayanthK007 committed:
added frontend testcases for subject component

JayanthK007 commented:
hi @seanlip for frontend test case failure i have added below code but E2E test case is failing there are many commits in this PR. Im not getting how to remove that oppia submodule.

for code coverage for front end. You can see below code.


  it('should enable the save button when input is entered', () => {
    spyOn(componentInstance, 'onChange');
    const event = {
      target: {
        value: 'science',
      },
    } as unknown as Event;

    let saveButton = fixture.debugElement.nativeElement.querySelector(
      '.e2e-test-save-changes-button'
    );
    expect(saveButton.disabled).toBeTrue();

    componentInstance.onInput(event);
    fixture.detectChanges();
    saveButton = fixture.debugElement.nativeElement.querySelector(
      '.e2e-test-save-changes-button'
    );
    expect(saveButton.disabled).toBeFalse();
  });

  it('should save input when focus is moved away', () => {
    spyOn(componentInstance, 'onChange');
    componentInstance.subjectInterests = [];
    componentInstance.subjectInterestInput = {
      nativeElement: {
        value: 'math',
      },
    } as ElementRef<HTMLInputElement>;

    componentInstance.onBlur();
    expect(componentInstance.subjectInterests).toEqual(['math']);
    expect(componentInstance.onChange).toHaveBeenCalledWith(['math']);
    expect(componentInstance.subjectInterestInput.nativeElement.value).toBe('');
  });
  
  for the functions which I have added Onblur and OnInput. Please can you please help me resolve this?

JayanthK007 committed:
resoved front end test case errors

JayanthK007 committed:
changed unit test case for subject interest component and made changes in logged in user for acceptance testing

JayanthK007 committed:
added front end testing

JayanthK007 committed:
Merge branch 'develop' into subject-interest

Ash-2k3 commented:
Hey @JayanthK007 I think the easy way to remove unwanted changes would be to bring back it to it's original state. In this case you can just delete the unwanted directory and commit the changes. Also before committing the changes, make sure to always see your changes using git commands like `git diff` and `git status`.

Also, I have assigned Sean for you, but for future references, feel free to leave a comment like `@{reviewers- id} PTAL` , Oppia bot will assign the reviewer for you then. Thanks.

JayanthK007 commented:
Hi @Ash-2k3, I'm facing trouble with front end testing as those are failing in CI. Give few more time, I'll try to solve those and again push my changes.

seanlip commented:
@JayanthK007 Re submodule, I don't really know how/why you added it. Perhaps close this PR and start afresh with a new branch?

For the frontend tests question, I don't know what your question is (it is not stated clearly), so I can't answer it.

JayanthK007 commented:
Hi @seanlip. I'll create a new pull request for this issue. The front end test coverage is failing. So I added test cases in `oppia\core\templates\pages\preferences-page\form-fields\subject-interests.component.spec.ts` After that it is failing 1 test case. I think i'll need help for this. Once i create new pull request. Can you please assign anyone to help?

seanlip commented:
@JayanthK007 If you want help on digging into an issue you're facing, standard procedure is to write a [debugging doc](https://github.com/oppia/oppia/wiki/Debugging-Docs) and then you can get help on GitHub Discussions. This will help us understand where you've gotten to and provide suggestions for next steps. Thanks!

JayanthK007 committed:
test files changed

JayanthK007 committed:
removed oppia folder

JayanthK007 committed:
Merge branch 'subject-interest' of https://github.com/JayanthK007/oppia into subject-interest

JayanthK007 commented:
Hi @seanlip I have removed oppia folder. For CI checks i'm getting E2E test failing for mobile. But it must work. Even frontend test cases I'm getting as some are not covered. So It is also failing. I have added comment regarding front end coverage for onBlur and onInput functions. Still it is failing can you please check and help me? If this 2 resolves you can merge this and close this.

JayanthK007 committed:
made changes for blur method in testing

JayanthK007 commented:
Hi @seanlip. I have solved E2E mobile testing. Only frontend testing is failing. I'll be trying to work on that. If you have any comments to make from the above code which I have added. It will help me solving that. Thanks!

seanlip commented:
@JayanthK007 See https://github.com/oppia/oppia/pull/21389#issuecomment-2566088206.

JayanthK007 commented:
Hi @seanlip sure. I'll write a debugging docs for frontend coverage testing. I'll try for 1 day if i wont be able to then I'll write the doc and ask help. Thanks..!

JayanthK007 committed:
added frontend coverage test cases

JayanthK007 committed:
added frontend coverage test cases

JayanthK007 committed:
lint checks done

JayanthK007 committed:
lint checking is done for front end

JayanthK007 committed:
lint check format done

JayanthK007 reviewed (commented):
Hi @seanlip I have added all test cases and all are passing in CI. You can review and merge to main branch. 

@imchristie @Ash-2k3 PTAL

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.spec.ts:
```diff
@@ -139,4 +139,48 @@ describe('Subject interests form field Component', () => {
+    componentInstance.onInput(inputEvent);
+    expect(componentInstance.validInput).toHaveBeenCalledWith('math');
+    expect(componentInstance.formCtrl.dirty).toBeTrue();
+    expect(componentInstance.onChange).toHaveBeenCalledWith([]);
```
Any expectation for the value of componentInstance.subjectInterests or componentInstance.subjectInterestInput.nativeElement.value?

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.spec.ts:
```diff
@@ -139,4 +139,48 @@ describe('Subject interests form field Component', () => {
+    componentInstance.onBlur();
+    expect(componentInstance.validInput).toHaveBeenCalledWith('math');
+    expect(componentInstance.subjectInterests).toEqual(['math']);
+    expect(componentInstance.subjectInterestInput.nativeElement.value).toBe('');
```
Can you explain why this is empty?

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.spec.ts:
```diff
@@ -139,4 +139,48 @@ describe('Subject interests form field Component', () => {
+      },
+    } as unknown as Event;
+    componentInstance.onInput(inputEvent);
+    expect(componentInstance.formCtrl.pristine).toBeTrue();
```
Need a pre-check to show that it was false before, otherwise you don't know that the cause of this being true was that the input is empty.

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.ts:
```diff
@@ -148,4 +148,26 @@ export class SubjectInterestsComponent implements ControlValueAccessor {
+
+    if (value && this.validInput(value)) {
+      this.subjectInterests.push(value);
+      this.subjectInterestInput.nativeElement.value = '';
```
Just for understanding -- what does nativeElement.value represent (on the page)?

Some comments here might be helpful.

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.ts:
```diff
@@ -148,4 +148,26 @@ export class SubjectInterestsComponent implements ControlValueAccessor {
+
+    if (!inputValue) {
+      this.formCtrl.markAsPristine();
+    } else if (this.validInput(inputValue)) {
```
Functions should start with a verb -- might want to rename to isValidInput.

seanlip commented on core/templates/pages/preferences-page/form-fields/subject-interests.component.ts:
```diff
@@ -148,4 +148,26 @@ export class SubjectInterestsComponent implements ControlValueAccessor {
+      this.formCtrl.markAsPristine();
+    } else if (this.validInput(inputValue)) {
+      this.formCtrl.markAsDirty();
+    }
```
If the input is not valid then what should we do? (Also this case doesn't seem to be tested.)

seanlip reviewed (changes requested):
Thanks @JayanthK007, sorry  for the late review. I left some comments, thanks.

Also, in the future, please ensure the Assignees list is correct, otherwise your review won't be picked up. You can check on the right hand side of the PR. Thanks.
